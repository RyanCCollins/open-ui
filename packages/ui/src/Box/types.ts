import * as React from 'react';
import Box from './';

export type Size = 'none' | 'small' | 'medium' | 'large' | 'xlarge';
export interface SizeObject { horizontal?: Size; vertical?: Size; };
export type Breakpoint = 'mobile' | 'tablet' | 'desktop';
export interface ResponsiveSize {
  mobile?: Size | SizeObject,
  tablet?: Size | SizeObject,
  desktop?: Size | SizeObject,
}
export type BoxSize =  'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge' | 'full';
export interface BoxSizeObject { horizontal?: BoxSize; vertical?: BoxSize; };
export interface FullObject { horizontal: boolean; vertical: boolean; };
export type Full = boolean | FullObject;
export type WrapOption = 'wrap' | 'wrap-reverse' | 'nowrap';
export type Rem = number;
export type Px = number;

export interface SizeStyle {
  height: string;
  width: string;
}

interface BoxProps extends React.HTMLProps<typeof Box> {
  alignItems?: 'flex-start' | 'center' | 'flex-end' | 'baseline' | 'stretch';
  flexDirection?: 'row' | 'column';
  justifyContent?: 'flex-start' | 'center' | 'flex-end' | 'space-between' | 'space-around';
  flexWrap?: boolean;
  reverse?: boolean;
  boxSize?: BoxSize | BoxSizeObject;
  pad?: Size | SizeObject;
  children?: JSX.Element;
  margin?: Size | SizeObject;
  backgroundColor?: string;
  backgroundImage?: string;
  full?: boolean | { horizontal?: boolean, vertical?: boolean };
  selectable?: boolean;
};

export default BoxProps;
