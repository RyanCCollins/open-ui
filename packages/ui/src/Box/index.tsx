import * as React from 'react';
import Component from './styles';
import BoxProps from './types';

export default function Box({
  children,
  ...rest,
}: BoxProps): JSX.Element {
  return (
    <Component {...rest}>
      {children || null}
    </Component>
  );
};
