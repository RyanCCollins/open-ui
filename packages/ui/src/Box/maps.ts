export const SIZE_MAP = {
  none: 0,
  small: 12,
  medium: 24,
  large: 48,
  xlarge: 96,
};

export const BOX_SIZE_MAP = {
  xxsmall: 48,
  xsmall: 96,
  small: 192,
  medium: 384,
  large: 576,
  xlarge: 720,
  xxlarge: 960,
};

export const BREAKPOINTS = {
  phone: 480,
  tablet: 768,
  desktop: 1024,
};
