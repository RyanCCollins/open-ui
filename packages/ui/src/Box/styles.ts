const styled = require('styled-components').default;
const { css } = require('styled-components');
import {
  calculateFlexWrap,
  calculateFullStyle,
  sizeToString,
  boxSizeToStyle,
} from './styleUtils';

export const BoxStyles = css`
  display: flex;
  background-color: ${({ backgroundColor }) => backgroundColor || 'transparent'};
  justify-content: ${({ justifyContent }) => justifyContent || 'flex-start'};
  align-items: ${({ alignItems }) => alignItems || 'flex-start'};
  flex-direction: ${({ flexDirection }) => flexDirection || 'column'};
  flex-wrap: ${({ flexWrap, reverse }) => calculateFlexWrap(flexWrap, reverse)};
  padding: ${({ pad }) => sizeToString(pad)};
  margin: ${({ margin }) => sizeToString(margin)};
  max-width: ${({ boxSize }) => boxSizeToStyle(boxSize).width};
  height: ${({ boxSize }) => boxSizeToStyle(boxSize).height};
  flex-basis: auto;
  min-height: ${({ full }) => calculateFullStyle(full, 'vh')};
  min-width: ${({ full }) => calculateFullStyle(full, 'vw')};
  cursor: ${({ selectable }) => selectable ? 'pointer' : 'inherit'};
`;

export default styled.div`
  ${BoxStyles}
`;
