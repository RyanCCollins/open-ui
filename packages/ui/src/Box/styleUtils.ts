const { css } = require('styled-components');
import { BREAKPOINTS } from './maps';
import {
  Size,
  SizeObject,
  ResponsiveSize,
  Breakpoint,
  BoxSize,
  BoxSizeObject,
  Full,
  WrapOption,
  SizeStyle,
  Rem,
  Px,
} from './types';
import { SIZE_MAP, BOX_SIZE_MAP } from './maps';

const rootRem: number = 16;
const remFromPX = (px: Px): Rem => (px / rootRem);

export function calculateFlexWrap(wrap?: boolean, reverse?: boolean): WrapOption {
  if (wrap && reverse) {
    return 'wrap-reverse';
  } else if (wrap && !reverse) {
    return 'wrap';
  } else {
    return 'nowrap';
  }
}

export function sizeToString(size: Size | SizeObject, smallSize: boolean = false): string {
  let returnVal;
  if (typeof size === 'string') {
    returnVal = `${remFromPX(SIZE_MAP[size])}rem`;
  } else if (typeof size === 'object') {
    const horizontal = size.horizontal || 'none';
    const vertical = size.vertical || 'none';
    returnVal = `${remFromPX(SIZE_MAP[vertical])}rem ${remFromPX(SIZE_MAP[horizontal])}rem`;
  } else {
    returnVal = `0rem`;
  }
  return returnVal;
}

function stringBoxStyle(size: BoxSize): SizeStyle {
  if (size === 'full') {
    return {
      width: '100vw',
      height: '100vh',
    };
  } else {
    return {
      width: size ? `${remFromPX(BOX_SIZE_MAP[size])}rem` : '',
      height: size ? `${remFromPX(BOX_SIZE_MAP[size])}rem` : '',
    };
  }
}

function objectBoxStyle(size: BoxSizeObject): SizeStyle {
  let width = 'auto';
  let height = 'auto';
  if (size.vertical) {
    height = size.vertical === 'full'
      ? '100vh'
      : `${remFromPX(BOX_SIZE_MAP[size.vertical])}rem`;
  }
  if (size.horizontal) {
    width = size.horizontal === 'full'
      ? '100vw'
      : `${remFromPX(BOX_SIZE_MAP[size.horizontal])}rem`;
  }
  return {
    width,
    height,
  };
}

export function boxSizeToStyle(size: BoxSize | BoxSizeObject): SizeStyle {
  if (typeof size === 'string') {
    return stringBoxStyle(size);
  } else if (typeof size === 'object') {
    return objectBoxStyle(size);
  } else {
    return { width: 'auto', height: 'auto' };
  }
}

export function calculateFullStyle(full: Full, postFix: 'vw' | 'vh'): string {
  if (typeof full === 'object') {
    if (postFix === 'vw') {
      return full.horizontal ? `100${postFix}` : 'auto';
    } else {
      return full.vertical ? `100${postFix}` : 'auto';
    }
  } else if (typeof full === 'boolean') {
    return full ? `100${postFix}` : 'auto';
  }
  return 'auto';
}

export const breakPointCss = (breakPoint: Breakpoint) => {
  const selector = breakPoint === 'desktop'
    ? 'min-width'
    : 'max-width';
  return css`
    @media screen and (${selector}: ${BREAKPOINTS.phone}) {
      padding: ${({ pad }) => sizeToString(pad)};
    }
  `;
};

export const breakPoints = (size: ResponsiveSize) => {
  if (size.desktop || size.mobile || size.tablet) {
    const css = Object.keys(size).map((key) =>
      breakPointCss(key as Breakpoint),
    ).join('; \n');
    return css;
  }
  return '';
};
