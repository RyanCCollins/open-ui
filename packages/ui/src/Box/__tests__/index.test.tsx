import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import * as React from 'react';
import Box from '../';
import props from './__mocks__/boxMocks.mock';

describe('<Box />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <Box>
        <div />
      </Box>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('should render with different props', () => {
    const wrapper = shallow(
      <Box {...props}>
        <div />
      </Box>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
