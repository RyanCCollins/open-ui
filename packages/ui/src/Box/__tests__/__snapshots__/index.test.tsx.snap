// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> should render with default props 1`] = `
<styled.div>
  <div />
</styled.div>
`;

exports[`<Box /> should render with different props 1`] = `
<styled.div
  pad="medium"
  size="large"
>
  <div />
</styled.div>
`;
